# Portfolio Project: Online Retail EDA with Python

## Project Overview  

This project explores an **online retail transactions dataset**, focusing on **data cleaning, exploratory data analysis (EDA), and deriving business insights**. The dataset contains information about customer purchases, including invoice details, product descriptions, quantities, prices, and customer IDs. 


### Objectives  
- Perform **Exploratory Data Analysis (EDA)** to identify key trends.  
- Analyze **sales performance, customer behavior, and popular products**.  
- Provide **data-driven recommendations** to optimize online retail strategies.  


### My Approach  
To tackle this project, I’ll start by **ETL (Extract, Transform, Load)** to clean and prepare the dataset. Then, I’ll conduct in-depth analysis to identify key trends and insights like **busiest sales periods, top-selling products, and high-value customers**. Let's dive in!  


## Dataset Overview


For this project, I'll be working with the **Online Retail** dataset, which contains transactional data from an online store between 2010 and 2011. The dataset is in a `.csv` file named **`online_retail.csv`**, and it includes details about purchases such as product descriptions, quantities, prices, timestamps, and customer IDs.  


### Data Columns  
The dataset consists of the following fields:  
- **InvoiceNo** – Unique invoice number for each transaction.  
- **StockCode** – Unique product identifier.  
- **Description** – Product name/description.  
- **Quantity** – Number of units purchased.  
- **InvoiceDate** – Timestamp of the transaction.  
- **UnitPrice** – Price per unit of the product.  
- **CustomerID** – Unique identifier for each customer.  
- **Country** – Country where the transaction took place.  


### My Approach  

To analyze this dataset effectively, I’ll break the process into key steps:  

1. **Load the data** into a Pandas DataFrame and inspect the first few rows.  
2. **Clean the dataset** by handling missing values and removing unnecessary data.  
3. **Explore basic statistics** to understand distributions and trends.  
4. **Visualize the data** using plots such as histograms, bar charts, and scatter plots.  
5. **Analyze sales trends** over time to identify peak sales periods.  
6. **Identify top-selling products and countries** based on quantity sold.  
7. **Detect anomalies or outliers** that may impact the analysis.  
8. **Summarize key findings** and insights from the data.  

Let's dive in and explore the dataset!  


# ETL

## 01. Load the data

Import the required libraries and load the dataset.

In [2]:
import pandas as pd
import matplotlib.pyplot as plt
import seaborn as sns

df = pd.read_csv("source/online_retail.csv", encoding="ISO-8859-1")  # We use encoding to avoid UnicodeDecodeError (or encoding="Windows-1252")

Explore and familiarize with the dataset

In [3]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 541909 entries, 0 to 541908
Data columns (total 8 columns):
 #   Column       Non-Null Count   Dtype  
---  ------       --------------   -----  
 0   InvoiceNo    541909 non-null  object 
 1   StockCode    541909 non-null  object 
 2   Description  540455 non-null  object 
 3   Quantity     541909 non-null  int64  
 4   InvoiceDate  541909 non-null  object 
 5   UnitPrice    541909 non-null  float64
 6   CustomerID   406829 non-null  float64
 7   Country      541909 non-null  object 
dtypes: float64(2), int64(1), object(5)
memory usage: 33.1+ MB


In [4]:
df.describe()

Quantity      UnitPrice     CustomerID
count  541909.000000  541909.000000  406829.000000
mean        9.552250       4.611114   15287.690570
std       218.081158      96.759853    1713.600303
min    -80995.000000  -11062.060000   12346.000000
25%         1.000000       1.250000   13953.000000
50%         3.000000       2.080000   15152.000000
75%        10.000000       4.130000   16791.000000
max     80995.000000   38970.000000   18287.000000

## 02. Clean the dataset

Now that we have identified the data types of each column and detected any missing (null) values, we have a clearer understanding of how to approach the ETL process.

Before proceeding, let's create a copy of the dataframe to preserve the original data in its unaltered state.

In [5]:
df_clean = df.copy()

### Type Casting

With the copy created, we will begin by modifying the data types of specific columns.  
In this case, we will convert the `Country`, `InvoiceNo`, and `StockCode` columns from the object type to the category type.  
This transformation will optimize memory usage and improve performance when handling these columns in Pandas.


In [6]:
df_clean['Country'] = df['Country'].astype('category')
df_clean['InvoiceNo'] = df['InvoiceNo'].astype('category')
df_clean['StockCode'] = df['StockCode'].astype('category')

# Ensure the data types where set correctly with: df_clean.info()

### Handling Missing Values

#### Handling missing `CustomerID` values

The dataset contains missing values in the `CustomerID` column, but these transactions are still valid purchases. Instead of dropping them or imputing arbitrary values (which could introduce bias), I will leave them as `NaN`.

 Why?
- Removing these rows would result in **loss of actual transaction data**.
- Imputing fake IDs would be **misleading**, as customer IDs are unique identifiers.
- Pandas and Matplotlib **handle NaN values gracefully** in most operations.

#### Handling missing `Description` values

The dataset contains null values in the `Description` column. Since these rows cannot be dropped without losing valuable data, we impute the missing descriptions using the corresponding `StockCode` values (which are complete and unique).

For that purpose, we follow this steps:
1. **Create a mapping dictionary** where each `StockCode` points to its correct `Description` (using only rows with non-null descriptions)
2. **Fill null values** by matching each missing `Description` with its `StockCode`'s known description

**Key Note**: If a `StockCode` has no valid description in the dataset, its `NaN` values will remain.

In [7]:
# Step 1: Map StockCode to Description (drop duplicates to ensure 1:1 mapping)
stock_to_desc = df_clean.dropna(subset=['Description']).drop_duplicates('StockCode').set_index('StockCode')['Description']

# Step 2: Fill NaN Descriptions using the mapped StockCode values
df_clean['Description'] = df_clean['Description'].fillna(df_clean['StockCode'].map(stock_to_desc))

After handling the preliminary missing values in the `Description` column, it's important to verify if any null values still remain. We will perform this check to ensure that all missing descriptions have been properly handled before moving forward with further analysis.

To do so, we'll check for any remaining nulls in the column.

In [8]:
# This will give us an updated count of the missing values in the 'Description' column
print(f'Original Description column null values: {df['Description'].isna().sum()}')
print(f'Updated Description column null values: {df_clean['Description'].isna().sum()}')

Original Description column null values: 1454
Updated Description column null values: 112


##### Imputing Remaining Null Values

After checking for null values, we found that 112 missing descriptions remain out of the initial 1,454 null values. To ensure we don't lose valuable transaction data, we will impute these remaining null values with the placeholder `'Unknown'`. This decision allows us to retain all rows in the dataset while clearly marking the transactions with missing descriptions.

In [9]:
df_clean['Description'] = df_clean['Description'].fillna('Unknown')

# To make sure this worked as intended: print(df_clean['Description'].isnull().sum())

By doing this, we preserve the full dataset while handling missing descriptions in a way that keeps the integrity of our analysis intact.

### String Processing

Before removing duplicates, we need to ensure that all truly identical rows are recognized as such by Pandas. To achieve this, we will standardize string formatting to eliminate inconsistencies.

We will focus on two key columns: `Description` and `Country`, as they contain string-type data.

`Description`: We will remove leading, trailing, and extra in-between whitespaces and standardize all text to uppercase for consistency.

`Country`: Similarly, we will trim unnecessary spaces and format country names in title case (first letter uppercase, the rest lowercase).

These transformations will help ensure that duplicate records are correctly identified and handled in the next step.

In [10]:
# Clean Description: Remove leading/trailing spaces, handle in-between extra spaces, and standardize to lowercase
df_clean['Description'] = df_clean['Description'].str.strip().str.replace(r'\s+', ' ', regex=True).str.upper()

# Clean Country: Remove leading/trailing spaces, handle in-between extra spaces, and title-case the country names
df_clean['Country'] = df_clean['Country'].str.strip().str.replace(r'\s+', ' ', regex=True).str.title()

### Removing Duplicates

Duplicated values can introduce bias and lead to incorrect insights, making it essential to handle them properly.

To begin, we will check for duplicate records in the dataset. Since individual columns may contain duplicate values, our focus will be on identifying and removing rows where all columns are identical.

For this, we will use Pandas' `.drop_duplicates()` method, which efficiently eliminates fully duplicated rows, ensuring data integrity for analysis.

In [11]:
print(f'Number of duplicate rows: {df_clean.duplicated().sum()}')

Number of duplicate rows: 5268


In [12]:
# Remove exact duplicate rows
df_clean = df_clean.drop_duplicates()

### Validating Negative Values

Based on the results of `df.describe()` in the *Load the Data* step, we identified negative values in the `Quantity` and `UnitPrice` columns. Since these values are not expected in a standard sales dataset, we will handle them systematically.

In [13]:
# Check for negative values in Quantity and UnitPrice
print(df_clean[df_clean['Quantity'] < 0].shape[0])
print(df_clean[df_clean['UnitPrice'] < 0].shape[0])

10587
2


Since the number of negative values in the `Quantity` column is significantly higher, we will address them first.

#### Analizing the negative values in `Quantity`

In [14]:
# Take a look at the negative values in Quantity and look for patterns
df_clean[df_clean['Quantity'] < 0].sample(20)

InvoiceNo StockCode                         Description  Quantity  \
4937     C536825     22617          BAKING SET SPACEBOY DESIGN        -1   
246779   C558727     22605           WOODEN CROQUET GARDEN SET        -1   
341273   C566741     23404          HOME SWEET HOME BLACKBOARD        -1   
96710    C544583         S                             SAMPLES        -1   
11518    C537251     22328   ROUND SNACK BOXES SET OF 4 FRUITS        -2   
140748   C548460     48138                  DOORMAT UNION FLAG        -2   
17134    C537684     22301        COFFEE MUG CAT + BIRD DESIGN        -1   
271255   C560653         M                              MANUAL        -1   
107374   C545436     20725             LUNCH BAG RED RETROSPOT        -1   
516373   C579886     21172                    PARTY METAL SIGN        -1   
160140   C550455     22382           LUNCH BAG SPACEBOY DESIGN        -1   
320761   C565089     22553              PLASTERS IN TIN SKULLS       -12   
75380    C542588     21730   GLASS STAR FROSTED T-LIGHT HOLDER        -2   
481165   C577346         M                              MANUAL        -1   
328579   C565789     22488  NATURAL SLATE RECTANGLE CHALKBOARD       -12   
297540   C562952     23111               PARISIENNE SEWING BOX        -1   
184926   C552720     22969        HOMEMADE JAM SCENTED CANDLES        -2   
356034   C567997     22776        SWEETHEART 3 TIER CAKE STAND        -1   
10204    C537234     22653                          BUTTON BOX       -20   
329373   C565843     22151           PLACE SETTING WHITE HEART        -1   

             InvoiceDate  UnitPrice  CustomerID         Country  
4937     12/2/2010 17:27       4.95     15384.0  United Kingdom  
246779    7/1/2011 14:35      10.95     12709.0         Germany  
341273   9/14/2011 14:55       4.95     15618.0  United Kingdom  
96710    2/21/2011 14:48      37.49         NaN  United Kingdom  
11518    12/6/2010 10:45       2.95         NaN  United Kingdom  
140748   3/31/2011 11:58       7.95     16801.0  United Kingdom  
17134    12/8/2010 10:18       2.55     14901.0  United Kingdom  
271255   7/20/2011 11:40     451.42     15802.0  United Kingdom  
107374    3/2/2011 15:52       1.65     13093.0  United Kingdom  
516373  11/30/2011 17:39       1.45     15676.0  United Kingdom  
160140   4/18/2011 13:02       1.65     13098.0  United Kingdom  
320761    9/1/2011 10:27       1.65     12645.0         Germany  
75380    1/28/2011 15:57       4.95     12997.0  United Kingdom  
481165  11/18/2011 15:25     530.25         NaN  United Kingdom  
328579     9/7/2011 9:07       1.65     16735.0  United Kingdom  
297540   8/11/2011 10:10      12.50     12749.0  United Kingdom  
184926    5/11/2011 9:49       1.45     18272.0  United Kingdom  
356034   9/23/2011 11:31       9.95     13097.0  United Kingdom  
10204     12/6/2010 9:40       1.95     16161.0  United Kingdom  
329373    9/7/2011 12:15       0.42     14606.0  United Kingdom

Upon examining the data, we can observe some patterns:  

1. Most `InvoiceNo` values associated with negative quantities begin with "C," which likely indicates a **credit transaction** (returns).  
2. Some `Description` values suggest special cases, such as "DAMAGED," "DISCOUNT," or "?".  

To better understand these cases, we will analyze how many unique `Description` values are associated with negative quantities.


In [15]:
negative_descriptions = df_clean[df_clean['Quantity'] < 0]['Description'].value_counts()
print(negative_descriptions.count())

2471


Given the large number of unique descriptions associated with negative quantities, we will focus on the most frequently occurring ones. This will help us identify common patterns and determine which descriptions may represent special cases that require specific handling.

In [16]:
print(negative_descriptions.head(30))  # Show the top 30 most frequent negative descriptions

Description
MANUAL                                 244
REGENCY CAKESTAND 3 TIER               180
POSTAGE                                126
CHECK                                  123
UNKNOWN                                 97
JAM MAKING SET WITH JARS                87
DISCOUNT                                77
SET OF 3 CAKE TINS PANTRY DESIGN        75
SAMPLES                                 61
DAMAGED                                 57
STRAWBERRY CERAMIC TRINKET BOX          54
ROSES REGENCY TEACUP AND SAUCER         54
RECIPE BOX PANTRY YELLOW DESIGN         47
DAMAGES                                 46
JUMBO BAG RED RETROSPOT                 44
LUNCH BAG RED RETROSPOT                 44
WOOD 2 DRAWER CABINET WHITE FINISH      43
RED RETROSPOT CAKE STAND                42
WHITE HANGING HEART T-LIGHT HOLDER      42
GREEN REGENCY TEACUP AND SAUCER         42
?                                       42
SMALL GLASS HEART TRINKET POT           40
SET OF 3 REGENCY CAKE TINS              37

#### Handling the negative values in `Quantity`

The output shows that some `Description` values represent regular products, while others indicate special cases (discounts, damaged products, or ambiguous values like "?"). We implement a unified classification system:

1. Categorize All Transactions

    Create a `TransactionType` column with three distinct labels:

    * **Return:** Transactions where `InvoiceNo` starts with "C" (credit notes).
    * **SpecialCase:** Transactions with descriptions matching predefined non-product terms (DISCOUNT, DAMAGED, SAMPLES, ?, etc.).
    * **Sale:** All other regular transactions.

2. Process Negative Quantities

    * **Returns:** Keep negatives (valid refund records).
    * **Special Cases:** Preserve original values (context-dependent).
    * **Sales:** Convert negatives to positives (assumed data entry errors).

**Note:** To distinguish legitimate negative quantities from data entry errors:

* Identified the top 30 most frequent descriptions for negative quantities

* Manually selected non-product terms (e.g., `DISCOUNT`, `DAMAGED`, `?`)

Resulting in the curated `special_case_list` used for classification.

In [17]:
# Predefined list of special cases for descriptions
special_case_list = [
    'DISCOUNT', 'DAMAGED', 'DAMAGES', 'SAMPLES', 'CHECK', 'MANUAL', 'POSTAGE', 
    'UNKNOWN', '?', 'AMAZON FEE', 'DOTCOM POSTAGE'
]

In [18]:
def classify_transaction(invoice_no, description):
    """Classify transaction as 'Return', 'SpecialCase', or 'Sale'."""
    if str(invoice_no).startswith('C'):
        return 'Return'
    elif description in special_case_list:
        return 'SpecialCase'
    else:
        return 'Sale'

# Use .map() efficiently by applying it on a tuple of (InvoiceNo, Description)
df_clean['TransactionType'] = list(map(classify_transaction, df_clean['InvoiceNo'], df_clean['Description']))

In [19]:
# Convert negatives to positives ONLY for regular Sales
df_clean.loc[
    (df_clean['TransactionType'] == 'Sale') & 
    (df_clean['Quantity'] < 0), 
    'Quantity'
] = df_clean['Quantity'].abs()  # Or: *= -1

To verify this approach worked as intended, we check if there are any negative `Quantity` values remaining that are neither classified as "Return" nor "SpecialCase".

In [20]:
print(f"Non Return or SpecialCase negative `Quantity` values: {df_clean[
    (df_clean['Quantity'] < 0) & 
    (~df_clean['TransactionType'].isin(['Return', 'SpecialCase']))
].shape[0]}")

Non Return or SpecialCase negative `Quantity` values: 0


Since the result is 0, it confirms that all negative `Quantity` values have been correctly handled according to our classification. We can now proceed to the next step.

#### Analizing and handling the negative values in `UnitPrice`

As observed earlier, only two transactions have negative `UnitPrice` values. Given their small number, we can inspect them directly as follows:

In [21]:
df_clean[df_clean['UnitPrice'] < 0]

InvoiceNo StockCode      Description  Quantity      InvoiceDate  \
299983   A563186         B  ADJUST BAD DEBT         1  8/12/2011 14:51   
299984   A563187         B  ADJUST BAD DEBT         1  8/12/2011 14:52   

        UnitPrice  CustomerID         Country TransactionType  
299983  -11062.06         NaN  United Kingdom            Sale  
299984  -11062.06         NaN  United Kingdom            Sale

Since they are labeled as "ADJUST BAD DEBT," these transactions appear to represent financial adjustments rather than product sales.

As these records are valid, we will keep them in the dataset and classify them under "SpecialCase" in the `TransactionType` column while preserving their negative values.


In [22]:
# Ensure "ADJUST BAD DEBT" transactions are marked as SpecialCase
df_clean.loc[df_clean['Description'] == 'ADJUST BAD DEBT', 'TransactionType'] = 'SpecialCase'

### Handling Outliers

The output from our initial exploration of `UnitPrice` shows that some transactions contain exceptionally high values. Upon further inspection, we found that several of these do not represent actual product sales but rather special cases, such as fees, postage, or adjustments.

1. Identifying High `UnitPrice` Transactions
To detect potential outliers, we examined the top transactions sorted by `UnitPrice`:

- Initially, we used `df_clean.nlargest(10, 'UnitPrice')` to inspect the highest values.

- Many high `UnitPrice` values corresponded to non-product transactions, which are **not currently classified as SpecialCases** (e.g., "DOTCOM POSTAGE", "THROW AWAY", "MOULDY, THROWN AWAY.").

- To refine our analysis, we iteratively excluded known non-product descriptions and re-ran the analysis to identify remaining cases.

In [23]:
df_clean.nlargest(10, 'UnitPrice')

InvoiceNo  StockCode      Description  Quantity      InvoiceDate  \
222681   C556445          M           MANUAL        -1  6/10/2011 15:31   
524602   C580605  AMAZONFEE       AMAZON FEE        -1  12/5/2011 11:36   
43702    C540117  AMAZONFEE       AMAZON FEE        -1    1/5/2011 9:55   
43703    C540118  AMAZONFEE       AMAZON FEE        -1    1/5/2011 9:57   
15016    C537630  AMAZONFEE       AMAZON FEE        -1  12/7/2010 15:04   
15017     537632  AMAZONFEE       AMAZON FEE         1  12/7/2010 15:08   
16356    C537651  AMAZONFEE       AMAZON FEE        -1  12/7/2010 15:49   
16232    C537644  AMAZONFEE       AMAZON FEE        -1  12/7/2010 15:34   
524601   C580604  AMAZONFEE       AMAZON FEE        -1  12/5/2011 11:35   
299982   A563185          B  ADJUST BAD DEBT         1  8/12/2011 14:50   

        UnitPrice  CustomerID         Country TransactionType  
222681   38970.00     15098.0  United Kingdom          Return  
524602   17836.46         NaN  United Kingdom          Return  
43702    16888.02         NaN  United Kingdom          Return  
43703    16453.71         NaN  United Kingdom          Return  
15016    13541.33         NaN  United Kingdom          Return  
15017    13541.33         NaN  United Kingdom     SpecialCase  
16356    13541.33         NaN  United Kingdom          Return  
16232    13474.79         NaN  United Kingdom          Return  
524601   11586.50         NaN  United Kingdom          Return  
299982   11062.06         NaN  United Kingdom     SpecialCase

2. Filtering Out Non-Product Cases
We applied a filtering step to exclude descriptions that represent fees, adjustments, or other special cases, ensuring that we focus on actual product-related outliers. The filtered descriptions include:

    "AMAZON FEE", "MANUAL", "DOTCOM POSTAGE", "BANK CHARGES", "ADJUST BAD DEBT", "POSTAGE", "DISCOUNT", "CRUK COMMISSION"

In [29]:
df_clean[~df_clean['Description'].isin(
    ["AMAZON FEE", "MANUAL", 'DOTCOM POSTAGE', 'BANK CHARGES', 'ADJUST BAD DEBT', 'POSTAGE', 'DISCOUNT', 'CRUK COMMISSION' 
    ])].nlargest(10, 'UnitPrice')

# 'DOTCOM POSTAGE', 'THROW AWAY', 'UNSALEABLE, DESTROYED.', 'MOULDY, THROWN AWAY.'

InvoiceNo StockCode                     Description  Quantity  \
222680    556444     22502  PICNIC BASKET WICKER 60 PIECES        60   
222682    556446     22502  PICNIC BASKET WICKER 60 PIECES         1   
242589   C558359         S                         SAMPLES        -1   
4989      536835     22655     VINTAGE RED KITCHEN CABINET         1   
32484     539080     22655     VINTAGE RED KITCHEN CABINET         1   
36165    C539438     22655     VINTAGE RED KITCHEN CABINET        -1   
51636     540647     22655     VINTAGE RED KITCHEN CABINET         1   
82768     543253     22655     VINTAGE RED KITCHEN CABINET         1   
87141    C543632     22655     VINTAGE RED KITCHEN CABINET        -1   
118769    546480     22656    VINTAGE BLUE KITCHEN CABINET         1   

             InvoiceDate  UnitPrice  CustomerID         Country  \
222680   6/10/2011 15:28      649.5     15098.0  United Kingdom   
222682   6/10/2011 15:33      649.5     15098.0  United Kingdom   
242589   6/28/2011 15:10      570.0         NaN  United Kingdom   
4989     12/2/2010 18:06      295.0     13145.0  United Kingdom   
32484    12/16/2010 8:41      295.0     16607.0  United Kingdom   
36165   12/17/2010 15:11      295.0     16607.0  United Kingdom   
51636    1/10/2011 14:57      295.0     17406.0  United Kingdom   
82768     2/4/2011 15:32      295.0     14842.0  United Kingdom   
87141    2/10/2011 16:22      295.0     14842.0  United Kingdom   
118769   3/14/2011 11:38      295.0     13452.0  United Kingdom   

       TransactionType  
222680            Sale  
222682            Sale  
242589          Return  
4989              Sale  
32484             Sale  
36165           Return  
51636             Sale  
82768             Sale  
87141           Return  
118769            Sale

To handle the identified outliers, we expand the `special_case_list` to include these descriptions and reuse the `classify_transaction` function, ensuring they are correctly categorized in the dataset.

In [25]:
# Expanding the special_case_list with newly identified non-product descriptions
special_case_list.extend([
    "DOTCOM POSTAGE", "THROW AWAY", "UNSALEABLE, DESTROYED.", "MOULDY, THROWN AWAY.",
    "AMAZON FEE", "MANUAL", "BANK CHARGES", "ADJUST BAD DEBT", "POSTAGE", "DISCOUNT", "CRUK COMMISSION"
])

# Reapplying the classification function to update TransactionType
df_clean['TransactionType'] = list(map(classify_transaction, df_clean['InvoiceNo'], df_clean['Description']))


Then, we can re-run the high `UnitPrice`analysis to confirm that all outliers are either properly classified or remain legitimate product sales.

By implementing this refined approach, we ensure that outlier removal is not arbitrary but instead data-driven, focusing on legitimate product transactions while flagging non-product entries as SpecialCases.

In [26]:
df_clean.nlargest(10, 'UnitPrice')

InvoiceNo  StockCode      Description  Quantity      InvoiceDate  \
222681   C556445          M           MANUAL        -1  6/10/2011 15:31   
524602   C580605  AMAZONFEE       AMAZON FEE        -1  12/5/2011 11:36   
43702    C540117  AMAZONFEE       AMAZON FEE        -1    1/5/2011 9:55   
43703    C540118  AMAZONFEE       AMAZON FEE        -1    1/5/2011 9:57   
15016    C537630  AMAZONFEE       AMAZON FEE        -1  12/7/2010 15:04   
15017     537632  AMAZONFEE       AMAZON FEE         1  12/7/2010 15:08   
16356    C537651  AMAZONFEE       AMAZON FEE        -1  12/7/2010 15:49   
16232    C537644  AMAZONFEE       AMAZON FEE        -1  12/7/2010 15:34   
524601   C580604  AMAZONFEE       AMAZON FEE        -1  12/5/2011 11:35   
299982   A563185          B  ADJUST BAD DEBT         1  8/12/2011 14:50   

        UnitPrice  CustomerID         Country TransactionType  
222681   38970.00     15098.0  United Kingdom          Return  
524602   17836.46         NaN  United Kingdom          Return  
43702    16888.02         NaN  United Kingdom          Return  
43703    16453.71         NaN  United Kingdom          Return  
15016    13541.33         NaN  United Kingdom          Return  
15017    13541.33         NaN  United Kingdom     SpecialCase  
16356    13541.33         NaN  United Kingdom          Return  
16232    13474.79         NaN  United Kingdom          Return  
524601   11586.50         NaN  United Kingdom          Return  
299982   11062.06         NaN  United Kingdom     SpecialCase

## 03. Explore basic statistics

Before diving into deeper analysis, it's essential to explore the dataset's basic statistics. This helps us understand key patterns, identify potential anomalies, and gain insights into the overall distribution of values.

By summarizing numerical and categorical data, we can:
- Detect **skewness or irregular distributions** in sales-related metrics.
- Identify **unexpected values**, such as extremely high or low prices and quantities.
- Assess **missing data patterns** that could impact further analysis.
- Ensure the dataset is **ready for meaningful insights** without significant biases.

Let's start by generating summary statistics and exploring key aspects of the dataset.